File: lib/day1.ml

  let rec find_first_last p2 v a b l = 
    if v then Printf.printf "Got %d %d %d:\n" (match a with None -> -1 | Some x -> x) (match b with None -> -1 | Some x -> x) (List.length l);
    match (a, b, l) with
      (None, _, []) | (_, None, []) -> failwith "Found no digits!"
      | (None, Some _, _ :: _) | (Some _, None, _ :: _) -> failwith "What the bleep!"
      | (Some x, Some y, []) -> (if v then Printf.printf " Found end with %d %d\n\n" x y; 10 * x + y)

      (* In the below, digits can overlap, so we'd better make sure we keep the overlapping letters. *)
      | (x, _, 'o' :: 'n' :: 'e' :: ll) when p2 -> (if v then Printf.printf " Found one\n"; find_first_last p2 v (if x = None then Some 1 else x) (Some 1) ('e' :: ll))
      | (x, _, 't' :: 'w' :: 'o' :: ll) when p2 -> (if v then Printf.printf " Found two\n"; find_first_last p2 v (if x = None then Some 2 else x) (Some 2) ('o' :: ll))
      | (x, _, 't' :: 'h' :: 'r' :: 'e' :: 'e' :: ll) when p2 -> (if v then Printf.printf " Found three\n"; find_first_last p2 v (if x = None then Some 3 else x) (Some 3) ('e' :: ll))
      | (x, _, 'f' :: 'o' :: 'u' :: 'r' :: ll) when p2 -> (if v then Printf.printf " Found four\n"; find_first_last p2 v (if x = None then Some 4 else x) (Some 4) ll)
      | (x, _, 'f' :: 'i' :: 'v' :: 'e' :: ll) when p2 -> (if v then Printf.printf " Found five\n"; find_first_last p2 v (if x = None then Some 5 else x) (Some 5) ('e' :: ll))
      | (x, _, 's' :: 'i' :: 'x' :: ll) when p2 -> (if v then Printf.printf " Found six\n"; find_first_last p2 v (if x = None then Some 6 else x) (Some 6) ll)
      | (x, _, 's' :: 'e' :: 'v' :: 'e' :: 'n' :: ll) when p2 -> (if v then Printf.printf " Found seven\n"; find_first_last p2 v (if x = None then Some 7 else x) (Some 7) ll)
      | (x, _, 'e' :: 'i' :: 'g' :: 'h' :: 't' :: ll) when p2 -> (if v then Printf.printf "Found eight\n"; find_first_last p2 v (if x = None then Some 8 else x) (Some 8) ('t' :: ll))
      | (x, _, 'n' :: 'i' :: 'n' :: 'e' :: ll) when p2 -> (if v then Printf.printf " Found nine\n"; find_first_last p2 v (if x = None then Some 9 else x) (Some 9) ('e' :: ll))

      (* Now match other stuff -- either digits or stuff we're ignoring. *)
      | (Some x, Some y, ch :: ll) -> (if v then Printf.printf " Found letter %c (with front %d and back %d)\n" ch x y; (find_first_last p2 v (Some x) (Some (if ch >= '0' && ch <= '9' then (int_of_char ch - int_of_char '0') else y)) ll))
      | (None, None, ch :: ll) -> (if v then Printf.printf " Found letter %c (with nothing)\n" ch; (let digit = match ch with | '0'..'9' -> Some (int_of_char ch - int_of_char '0') | _ -> None in find_first_last p2 v digit digit ll))

 let process_line part_two verbose line = 
  find_first_last part_two verbose None None (List.of_seq (String.to_seq line))

  
let day1 input_lines part_two verbose _ = 
  let line_list = (List.map (process_line part_two verbose) (List.filter (fun l -> String.length l > 0) input_lines)) in 
    let sum = List.fold_left ( + ) 0 line_list in
      print_endline (string_of_int sum)

File: python/day1.py
#!/usr/bin/env python3

import re
import sys, os

import adventday

class Day1(adventday.AdventDay):
    def calc_calibration(self, line):
        first = None
        last = None
        line = line.strip()
        if len(line) == 0:
            return 0
        for i in range(len(line)):
            if line[i:].startswith('0') or line[i:].startswith('1') or line[i:].startswith('2') or line[i:].startswith('3') or line[i:].startswith('4') or line[i:].startswith('5') or line[i:].startswith('6') or line[i:].startswith('7') or line[i:].startswith('8') or line[i:].startswith('9'):
                last = ord(line[i]) - ord('0')
            elif self.args.part_two:
                if line[i:].startswith('one'):
                    last = 1
                elif line[i:].startswith('two'):
                    last = 2
                elif line[i:].startswith('three'):
                    last = 3
                elif line[i:].startswith('four'):
                    last = 4
                elif line[i:].startswith('five'):
                    last = 5
                elif line[i:].startswith('six'):
                    last = 6
                elif line[i:].startswith('seven'):
                    last = 7
                elif line[i:].startswith('eight'):
                    last = 8
                elif line[i:].startswith('nine'):
                    last = 9

            if first is None and last is not None:
                first = last
        assert first is not None and last is not None, F'Eek! Line {line} gives first={first} and last={last}!'
        return 10 * first + last

    def part_one(self):
        return sum(self.calc_calibration(x) for x in self.lines)

    def part_two(self):
        return self.part_one()